### ISSUE TITLE
Benutzerdefinierte Header um CORS block zu umgehen (Access-Control-Allow-Origin:)

### ISSUE THREAD

**Issue by backlink-market** [Zefau/ioBroker.jarvis#2415]
Hi,

I kann meine Hikvision Cameras nicht in Jarvis einbinden, da die Bilder durch die Cross-Origin Resource Sharing (CORS) Policy vom Browser geblockt werden.
Um das zu umgehen muss der "Access-Control-Allow-Origin: *" Header hinzugefügt werden.
Kann man das implementieren?

Vielen Dank

**Comment by Zefau** [comment]
Kannst du die IP nicht whitelisten bzw. Zugriff von allen Hosts / IPs erlauben?

![image](https://github.com/Zefau/ioBroker.jarvis/assets/14368811/b685ad09-6dee-4b94-9bac-f55d2997a805)

**Comment by backlink-market** [comment]
gibts dazu was neues?

**Comment by Zefau** [comment]
Verstehe es noch nicht ganz. Du nutzt ein iFrame mit dem du die Kamera aufrufst? Ich weiß nicht, wo genau ich den CORS Header einbauen soll. Der muss ja in den Request an deine Kamera?

**Comment by backlink-market** [comment]
Genau, wenn das DisplayImage die Kamera URL aufruft musst du den mitsenden.

**Comment by Zefau** [comment]
Also DisplayImage ist ja einfach nur ein `<img .. />` HTML-Tag. Da kann ich `crossorigin` aufnehmen, siehe https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/crossorigin. Können wir mit der nächsten Alpha probieren.

**Comment by backlink-market** [comment]
nein, nicht als html img tag sondern im header mit "Access-Control-Allow-Origin: *" 
siehe  https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

**Comment by Zefau** [comment]
Das ergibt aber keinen Sinn. Wo soll das rein? In welchen Header? Der Request von jarvis an deine Kamera geht von dem HTML-Tag `<img />` aufgrund der Angabe in `src` aus. jarvis schickt sonst keinen Request außer das.

**Comment by backlink-market** [comment]
ok, dann probieren wir es so.

**Comment by Zefau** [comment]
Dieser Feature Request wurde mit `v3.2.0-alpha.43` implementiert. Bitte bestätigen und auf Fehler prüfen. Wenn alles in Ordnung ist, gerne das Issue schließen.<br /><hr />This feature request has been implemented with `v3.2.0-alpha.43`. Please verify and test the feature for any bugs. If everything works as expected, you may close the issue.

**Comment by backlink-market** [comment]
wo bekomm ich die version v3.2.0-alpha.43?
Auf der release page https://github.com/Zefau/ioBroker.jarvis/releases ist nur bis v3.2.0-alpha.41

**Comment by mcuiobroker** [comment]
Adapter ->Expert-Mode -> Katze ->github -> jarvis

**Comment by backlink-market** [comment]
sorry, aber die version ist total buggy. Kann weder meine oberfläche aufrufen, noch das layout:
https://prnt.sc/G47baAA_MuhA

**Comment by Zefau** [comment]
Wie sieht denn die URL aus?

Wenn du dir lokal eine HTML-Datei erstellst und via `<img />` versuchst das Kamera-Bild zu laden, bekommst du den Fehler dann auch?

```
<img src=" ... " />
```
oder
```
<img src=" ... " crossorigin="anonymous" />
```
oder

```
<img src=" ... " crossorigin="use-credentials" />
```

https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_enabled_image
https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-crossorigin
https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-crossorigin

**Comment by backlink-market** [comment]
no, none of these version work. I also added a .php file on a local server to add the cors header:

But this does not work, too. When I open the URL in the browser, it works, But not as img nor as iframe.

```<?php header('Access-Control-Allow-Origin: *'); ?>
<!doctype html>
<html>
<head>
<meta charset="utf-8">
<title>Unbenanntes Dokument</title>
</head>

<body>
<h2>1.</h2> <img src="http://username:password@192.168.1.5/ISAPI/Streaming/Channels/102/picture" />


</body>
</html>

**Comment by Zefau** [comment]
Hmm, dann weiß ich nicht wie ich das in Jarvis lösen soll :-/ Ist ja nur die <IMG Tag auf den ich Einfluss habe beim Request. Mehr geht ja von Jarvis nicht raus

**Comment by Zefau** [comment]
I guess the result is the same for every browser?

**Comment by Zefau** [comment]
> http://username:password@192.168.1.5/ISAPI/Streaming/Channels/102/picture

This will not work anymore with newer browsers. You have to base64 the credentials.
